File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/barrel/MilkingCategory.java
package wraith.fabricaeexnihilo.compatibility.rei.barrel;

import me.shedaniel.math.Point;
import me.shedaniel.math.Rectangle;
import me.shedaniel.rei.api.client.gui.Renderer;
import me.shedaniel.rei.api.client.gui.widgets.Widget;
import me.shedaniel.rei.api.client.gui.widgets.Widgets;
import me.shedaniel.rei.api.client.registry.display.DisplayCategory;
import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.util.EntryStacks;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.GlyphWidget;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;

import java.util.ArrayList;
import java.util.List;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

public class MilkingCategory implements DisplayCategory<MilkingDisplay> {

    public static final Identifier ARROW = id("textures/gui/rei/glyphs.png");
    public static final int MARGIN = 6;
    public static final int HEIGHT = 3 * 18 + MARGIN * 2;
    public static final int ARROW_Y = HEIGHT / 2;
    public static final int BARRELS_Y = ARROW_Y;
    public static final int ABOVE_Y = BARRELS_Y - 18;
    public static final int OUTPUT_Y = ARROW_Y;
    public static final int WIDTH = 8 * 18 + MARGIN * 2;
    public static final int ARROW_X = WIDTH / 2 - 9;
    public static final int BARRELS_X = ARROW_X - 18;
    public static final int ABOVE_X = BARRELS_X;
    public static final int OUTPUT_X = ARROW_X + 18;
    private final ItemStack icon;
    private final String name;

    public MilkingCategory(ItemStack icon, String name) {
        this.icon = icon;
        this.name = name;
    }

    @Override
    public CategoryIdentifier<? extends MilkingDisplay> getCategoryIdentifier() {
        return PluginEntry.MILKING;
    }

    @Override
    public int getDisplayHeight() {
        return HEIGHT;
    }

    @Override
    public int getDisplayWidth(MilkingDisplay display) {
        return WIDTH;
    }

    @Override
    public Renderer getIcon() {
        return EntryStacks.of(icon);
    }

    @Override
    public Text getTitle() {
        return Text.translatable(name);
    }

    @Override
    public List<Widget> setupDisplay(MilkingDisplay display, Rectangle bounds) {
        var widgets = new ArrayList<Widget>();
        widgets.add(Widgets.createRecipeBase(bounds));

        widgets.add(new GlyphWidget(bounds, bounds.getMinX() + ARROW_X, bounds.getMinY() + ARROW_Y, 16, 16, ARROW, 0, 0));

        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + ABOVE_X, bounds.getMinY() + ABOVE_Y)).entries(display.entity).markInput());
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + BARRELS_X, bounds.getMinY() + BARRELS_Y)).entries(PluginEntry.BARRELS.get()));
        widgets.add(Widgets.createSlot(new Point(bounds.getMinX() + OUTPUT_X, bounds.getMinY() + OUTPUT_Y)).entries(display.result).markOutput());

        //TODO: some day figure out how to render a little entity over the barrel instead of an egg :/ (steal code from witchwater entity recipe?)
//        (display.recipe.entity.flatten().firstOrNull())?.let{
//            widgets.add(EntityWidget(bounds.getMinX(), bounds.getMinY(), 36, 36, it))
//        }

        var text = Widgets.createLabel(new Point(0, 0), Text.literal(String.valueOf(display.amount)));
        text.setPoint(new Point(bounds.getMaxX() - MARGIN - text.getBounds().getMaxX(), bounds.getMinY() + MARGIN));
        widgets.add(text);

        return widgets;
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/barrel/MilkingCategory.kt
package exnihilofabrico.compatibility.rei.barrel

import exnihilofabrico.compatibility.rei.GlyphWidget
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.id
import exnihilofabrico.util.asREIEntry
import me.shedaniel.math.api.Point
import me.shedaniel.math.api.Rectangle
import me.shedaniel.rei.api.RecipeCategory
import me.shedaniel.rei.gui.widget.EntryWidget
import me.shedaniel.rei.gui.widget.LabelWidget
import me.shedaniel.rei.gui.widget.RecipeBaseWidget
import me.shedaniel.rei.gui.widget.Widget
import net.minecraft.item.Items
import java.util.function.Supplier

class MilkingCategory: RecipeCategory<MilkingDisplay> {

    override fun getIdentifier() = PluginEntry.MILKING
    override fun getLogo() = Items.MILK_BUCKET.asREIEntry()
    override fun getCategoryName() = "Barrel Milking"


    override fun getDisplayHeight() = HEIGHT
    override fun getDisplayWidth(display: MilkingDisplay) = WIDTH

    override fun setupDisplay(displaySupplier: Supplier<MilkingDisplay>, bounds: Rectangle): MutableList<Widget> {
        val display = displaySupplier.get()
        val widgets = mutableListOf<Widget>(RecipeBaseWidget(bounds))

        val arrow = GlyphWidget(bounds, bounds.minX + ARROW_X, bounds.minY + ARROW_Y, ARROW_WIDTH, ARROW_HEIGHT, ARROW, ARROW_U, ARROW_V)
        widgets.add(arrow)

        val eggs = display.inputEntries[0]
        val barrels = display.inputEntries[1]
        val outputs = display.outputEntries

        widgets.add(EntryWidget.create(bounds.minX + ABOVE_X, bounds.minY + ABOVE_Y).entries(eggs))
        widgets.add(EntryWidget.create(bounds.minX + BARRELS_X, bounds.minY + BARRELS_Y).entries(barrels))
        widgets.add(EntryWidget.create(bounds.minX + OUTPUT_X, bounds.minY + OUTPUT_Y).entries(outputs))

        // TODO some day figure out how to render a little entity over the barrel instead of an egg :/
//        (display.recipe.entity.flatten().firstOrNull())?.let{
//            widgets.add(EntityWidget(bounds.minX, bounds.minY, 36, 36, it))
//        }


        val text =  LabelWidget(0, 0, display.recipe.result.localizeAmount())
        text.position = Point(bounds.maxX - MARGIN - text.bounds.maxX, bounds.minY + MARGIN)
        widgets.add(text)


        return widgets
    }

    companion object {
        val ARROW = id("textures/gui/rei/glyphs.png")

        val MARGIN = 6

        val WIDTH = 8*18 + MARGIN*2
        val HEIGHT = 2*18 + MARGIN*2

        val ARROW_X = WIDTH/2 - 9
        val ARROW_Y = HEIGHT/2

        val BARRELS_X = ARROW_X - 18
        val BARRELS_Y = ARROW_Y

        val ABOVE_X = BARRELS_X
        val ABOVE_Y = BARRELS_Y - 18

        val OUTPUT_X = ARROW_X + 18
        val OUTPUT_Y = ARROW_Y

        val ARROW_WIDTH = 16
        val ARROW_HEIGHT= 16
        val ARROW_U = 0
        val ARROW_V= 0
    }

}